Introduce Pattern Library test pages

andy-armstrong opened the pull request:
https://openedx.atlassian.net/browse/FEDX-93

This change renames the current LMS and Studio CSS files as (lms/studio)-main-v1.css, and introduces new Pattern Library versions at (lms/studio)-main-v2.css (the same was done for the RTL versions). A page template can now choose to use the Pattern Library by including this at the top of the Mako template:

    ## Override the default styles_version to the Pattern Library version (version 2)
    <%!
        styles_version = 2
    %>

For now these pages have no styling, but that will be introduced with https://openedx.atlassian.net/browse/FEDX-118. You can see the current state of the LMS test pages here:

https://andy-armstrong.sandbox.edx.org/template/ux/reference/pattern-library-test.html

New developer-only test pages are available locally at:

http://localhost:8000/template/ux/reference/pattern-library-test.html
http://localhost:8001/template/ux/reference/pattern-library-test.html

@AlasdairSwan @dan-f please review.

 @saleem-latif @douglashall @mattdrayer I'd love your feedback both from a comprehensive theming perspective as from how to build new pattern library pages.

FYI @cahrens @bjacobel @dsjen @tobz @clrux

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

AlasdairSwan commented:
Looking good @andy-armstrong. The only thing I am seeing is a logo image file not loading in lms. Are you seeing this locally? 
![screen shot 2016-03-25 at 3 52 23 pm](https://cloud.githubusercontent.com/assets/2052185/14052603/d0b87526-f2a1-11e5-9184-b8aed74435a5.png)

andy-armstrong commented:
@AlasdairSwan Yep, I'm seeing the same errors. I'll look into them...

andy-armstrong commented:
@AlasdairSwan This image is coming from the footer, so I don't believe my changes have caused this. It looks like a problem/misconfiguration with comprehensive theming. Interestingly the rest of the page is in the Open edX theme, so I'm not sure why it is trying to load the edx.org theme's logo.

andy-armstrong commented:
@AlasdairSwan I've made a couple of improvements:
 - I've made it so that each page can specify an exact filename for their main CSS file. My thinking is that we can then have custom CSS per page, e.g. a very stripped down CSS for the login page.
 - I've removed the need to register single CSS groups for individual files, so that it is possible to include any CSS file that a page wants.

dan-f commented on cms/static/sass/_config.scss:
```diff
@@ -0,0 +1,11 @@
+// ------------------------------
+// Open edX Certificates: Config
```
This looks like a copy-paste error

dan-f commented on cms/static/sass/_lib.scss:
```diff
@@ -0,0 +1,9 @@
+// ------------------------------
+// Open edX Certificates: Main Style Compile
```
Copy paste error?

dan-f commented on cms/static/sass/_ltr.scss:
```diff
@@ -0,0 +1,24 @@
+// ------------------------------
+// Open edX Certificates: Main Style Compile
```
Copy paste error?

dan-f commented on cms/static/sass/_rtl.scss:
```diff
@@ -0,0 +1,24 @@
+// ------------------------------
+// Open edX Certificates: Main Style Compile
```
Copy paste error?

dan-f commented on cms/static/sass/_lib.scss:
```diff
@@ -0,0 +1,9 @@
+// About: third party libraries and dependencies import
+
+
+@import '../../../node_modules/edx-pattern-library/node_modules/bourbon/app/assets/stylesheets/bourbon';
```
I'm worried about reaching into the Pattern Library's `node_modules` directory. I think we should encapsulate this within a pattern library SASS module.

andy-armstrong replied:
@dan-f This is a really good point. I asked Brian the same thing when he first imported the Pattern Library, and he said it wasn't possible. You can see his response here:

https://github.com/edx/edx-platform/pull/11307#discussion-diff-51042710

Reading that back, I'm not sure I understand it. It seems like it should be possible to make it work, but I don't want to hold up this work while we investigate other options.

dan-f replied:
I'm not sure that I understand @talbs's comment (maybe he'll kindly 'splain to me here... :))

> 1. We don't know where an individual app will store its npm-installed libraries, which includes any libs pulled in as UXPL dependencies. So a path made in the UXPL assets wouldn't be able to account for that until we can pass variables into an @import path (see sass/sass#279).

edx-platform consumes UXPL. UXPL knows how npm is configured, hence the local path from the UXPL project root to its `node_modules` directory, and therefore `_rtl.scss` can safely encapsulate that. edx-platform knows _its_ npm configuration, hence it knows where `_rtl.scss` is installed. I am genuinely confused as to what's preventing us from importing the `_rtl.scss` (or any other) partial.

> 2. The `$layout-direction: ltr;`, `$grid-direction-default: ltr;`, and `$grid-direction-reversed: ltr;` variables need to be set specifically for an app's main `*-ltr.scss` and `*--rtl.scss` compile files. I had originally had this logic in the UXPL using the `@if` + `@import` method in the UXPL that we removed recently (edx/ux-pattern-library#263).

It makes total sense to me for us to define our own SASS variables in this way.

dan-f commented on cms/static/sass/_ltr.scss:
```diff
@@ -0,0 +1,24 @@
+// ----------------------------
+// #LIB
+// ----------------------------
+@import '../../../node_modules/edx-pattern-library/node_modules/bi-app-sass/bi-app/bi-app-ltr';
```
Same comment as above regarding encapsulation.

dan-f commented on cms/static/sass/_rtl.scss:
```diff
@@ -0,0 +1,24 @@
+// ----------------------------
+// #LIB
+// ----------------------------
+@import '../../../node_modules/edx-pattern-library/node_modules/bi-app-sass/bi-app/bi-app-rtl';
```
(encapsulation note)

dan-f commented on common/djangoapps/pipeline_mako/__init__.py:
```diff
@@ -24,6 +24,28 @@ def compressed_css(package_name, raw=False):
+    single CSS file.
+
+    Args:
+        package_name: The name of the CSS package to be
```
to be...?

dan-f commented on lms/static/sass/_config.scss:
```diff
@@ -0,0 +1,11 @@
+// ------------------------------
+// Open edX Certificates: Config
```
copy paste

dan-f commented on lms/static/sass/_lib.scss:
```diff
@@ -0,0 +1,9 @@
+// ------------------------------
+// Open edX Certificates: Main Style Compile
```
:)

dan-f commented:
I'm curious about the design decision behind `v1`/`v2` etc. Is `v2` the full edX LMS/CMS stylesheet, covering everything that the old `v1` CSS did? Do we expect every page wishing to use the pattern library to import the `v2` style? Should our base stylesheets grow linearly in style support such that `v[i] ⊂ v[i+1]` in terms of styles? I remember us discussing a modular design in which each webpage could declare the styles it needs by using its own `.scss` file which imports only the needed modules from the pattern library. Is that still the goal, and these `main-v2` files are the bare minimum bootstrapping styles?

andy-armstrong commented:
@dan-f Thanks for the thorough feedback, and great question about the v1/v2. My thought process was that there will be a standard vanilla style that most simple new pages will adopt. New features should indeed have their own custom Sass file that brings in only what they need, and which won't unnecessarily bloat the shared main styles. Does that make sense?

dan-f commented:
Thanks for the clarification. That makes sense, so :+1: on the conceptual level :)

I've got some follow up questions. Does this PR reintroduce that vanilla style as implemented by the Pattern Library? Are the APIs (class names, etc) consistent between v1 and v2, or will developers need to use new APIs to get their UI styled by the pattern library? If the latter, then I wonder if the vanilla style template, which is the same as `pattern-library.scss`, should be called something like `pattern-library-main.scss` so as to differentiate between legacy LMS/CMS styles and the new Pattern Library ones. Or should we even just import the pattern library directly?

andy-armstrong commented:
@dan-f Pattern Library class names etc are not guaranteed to be consistent with the v1 styles, so you have to code for one or the other. My original implementation named the new file pattern-library-main.css, but @AlasdairSwan pointed out that it is just an implementation detail. We considered many other names, but we both liked v2 as being indicative of an incompatible switch where v2 is better to use than v1. 

Note that v2 is currently only the Pattern Library styles, but it will very shortly have common implementations for the header and footer, and other globally useful styles. We definitely can't just import the pattern library CSS directly as it doesn't have the appropriate theming variables applied.

andy-armstrong commented:
@AlasdairSwan @dan-f I've addressed the feedback (except for the way that the Pattern Library is imported which is beyond the scope of this change), and squashed the changes. Please re-review.

FYI, you can see @AlasdairSwan's work on styling the header for this demo page here:

https://andy-armstrong.sandbox.edx.org/template/ux/reference/pattern-library-test.html

dan-f commented:
I now agree with the naming, given that those styles will eventually include more than just the pattern library. I'm still curious about the import issue - what makes it out of scope, and have you tried importing the partials from the pattern library?

andy-armstrong commented:
@dan-f Thanks for the push. I've read up on Sass import paths, and I now think there's a much easier way for this to be managed. See the related Pattern Library PR - https://github.com/edx/ux-pattern-library/pull/313. Is this more what you had in mind? It has greatly simplified the amount of boilerplate, and has kept all the knowledge of third party libraries inside the Pattern Library where it belongs.

dan-f commented:
@andy-armstrong that's awesome, and just what I was hoping for! :+1000:

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
jenkins run bokchoy

andy-armstrong commented:
@AlasdairSwan @dan-f I've fixed the test failures so I think this is ready to go. Please re-review.

mattdrayer commented:
@malikshahzad228 FYI

saleem-latif commented:
@andy-armstrong the changes look good from theming perspective. I tried some pattern library overrides (like fonts overrides) and they seem to be working on themes.

AlasdairSwan commented on cms/templates/ux/reference/pattern-library-test.html:
```diff
@@ -0,0 +1,27 @@
+
+            <p class="alert-copy">
+                Interesting pattern library content to come...
+            </p>
```
Nitpick because it's only a test page but the alert title should be different from the copy, otherwise a screenreader user will hear the same thing twice. In this case I don't think we beed the title.

AlasdairSwan commented on lms/templates/ux/reference/pattern-library-test.html:
```diff
@@ -0,0 +1,30 @@
+
+            <p class="alert-copy">
+                Interesting pattern library content to come...
+            </p>
```
Nitpick because it's only a test page but the alert title should be different from the copy, otherwise a screenreader user will hear the same thing twice. In this case I don't think we beed the title.

AlasdairSwan commented:
Looking good @andy-armstrong. Nitpicks around the test pages, but once addressed :+1:

andy-armstrong commented:
jenkins run bokchoy

dan-f commented:
:+1:

douglashall commented:
FYI @afzaledx @hasnain-naveed
